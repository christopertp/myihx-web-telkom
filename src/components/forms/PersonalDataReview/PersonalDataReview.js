import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { Field } from 'redux-form';
import Button from '../../elements/Button';
import Spinner from '../../elements/Spinner';
import TextField from '../../fields/Text';
import styles from './styles.scoped.css';

export default function PersonalDataReview(props) {
  const { handleSubmit, initialize, invalid, onRetake } = props;
  const { page, subpage } = useParams();
  const { form, isLoading, message } = useSelector(s => s.personalData);
  const labels = {
    'e-ktp': 'e-KTP',
    'ktp': 'KTP',
    'passport': 'passport',
    'sim': 'driver\'s license',
  };

  const idNumberProps = {
    label: `${labels[page].toUpperCase()} NUMBER`,
    maxLength: 16,
    placeholder: `Enter ${labels[page]} number`,
  };

  const nameProps = {
    label: 'NAME',
    placeholder: `Enter name according to ${labels[page]}`,
  };

  const motherNameProps = {
    label: 'MOTHER\'S MAIDEN NAME',
    placeholder: 'Enter mother\'s maiden name',
  };

  const dateOfBirthProps = {
    label: 'DATE OF BIRTH',
    max: new Date().toISOString().split('T')[0],
    placeholder: 'Enter date of birth',
    type: 'date',
  };

  const validate = v => !v ? 'Harus diisi' : '';

  useEffect(() => {
    if (subpage === 'edit') {
      const { dateOfBirth, idNumber, motherName } = form;
      initialize({ dateOfBirth: page === 'ktp' ? dateOfBirth : undefined, idNumber, motherName });
    }
  }, []);

  return (
    <form className={styles.root} onSubmit={handleSubmit}>
      <Field component={TextField} inputProps={idNumberProps} name="idNumber" />
      <Field component={TextField} inputProps={nameProps} name="name" />
      <Field component={TextField} inputProps={motherNameProps} name="motherName" />
      {page === 'ktp' && <Field component={TextField} inputProps={dateOfBirthProps} name="dateOfBirth" validate={validate} />}
      {message && <small className={styles.error}>{message}</small>}
      <div>
        <Button disabled={isLoading} onClick={onRetake} variant="bordered">Reupload</Button>
        <Button disabled={invalid || isLoading} type="submit">
          {isLoading ? <Spinner /> : 'Confirm'}
        </Button>
      </div>
    </form>
  );
}

PersonalDataReview.defaultProps = {
  handleSubmit: () => {},
  initialize: () => {},
  invalid: false,
  onRetake: () => {},
};

PersonalDataReview.propTypes = {
  handleSubmit: PropTypes.func,
  initialize: PropTypes.func,
  invalid: PropTypes.bool,
  onRetake: PropTypes.func,
};
