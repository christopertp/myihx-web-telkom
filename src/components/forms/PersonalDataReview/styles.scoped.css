.root > label:not(:first-of-type) {
  margin-top: 1.25rem;
}

.root > div:last-child {
  display: flex;
  justify-content: space-between;
  margin-top: 2rem;
}

.root > div:last-child > * {
  flex: 1
}

.root > div:last-child > *:first-child {
  margin-right: 1rem;
}

.error {
  color: #ee4c24;
  display: block;
  font-size: 0.75rem;
  margin-top: 1rem;
}
