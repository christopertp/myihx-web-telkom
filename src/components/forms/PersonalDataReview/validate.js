export default function validate(values) {
  const { idNumber, motherName, name } = values;

  return {
    idNumber: !/^[0-9\s]+$/.test(idNumber) ? 'Harus diisi dengan angka' : '',
    motherName: !/^[a-zA-Z\s]+$/.test(motherName) ? 'Harus diisi dengan alfabet' : '',
    name: !/^[a-zA-Z\s]+$/.test(name) ? 'Harus diisi dengan alfabet' : '',
  };
}
