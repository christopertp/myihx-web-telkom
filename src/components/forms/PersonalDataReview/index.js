import { reduxForm } from 'redux-form';
import PersonalDataReview from './PersonalDataReview';
import validate from './validate';

export default reduxForm({
  initialValues: {
    idNumber: '',
    motherName: '',
    name: '',
  },
  form: 'personalDataReview',
  validate,
})(PersonalDataReview);
