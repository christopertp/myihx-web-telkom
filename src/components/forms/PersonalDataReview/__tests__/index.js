import React from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import Spinner from '../../../elements/Spinner';
import PersonalDataReview from '../PersonalDataReview';


useParams.mockImplementation(() => ({ page: 'e-ktp' }));

useSelector.mockImplementation(fn => {
  fn({});
  return { form: '' };
});

describe('src/components/forms/PersonalDataReview', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<PersonalDataReview />);
    expect(tree).toMatchSnapshot();
    PersonalDataReview.defaultProps.handleSubmit();
    PersonalDataReview.defaultProps.initialize();
    PersonalDataReview.defaultProps.onRetake();
  });

  test('loading & error state', () => {
    const initialize = jest.fn();
    const props = { ...PersonalDataReview.defaultProps, initialize };
    useParams.mockImplementation(() => ({ page: 'e-ktp', subpage: 'edit' }));
    useSelector.mockImplementationOnce(() => ({ form: '', isLoading: true, message: 'error' }));

    const result = PersonalDataReview(props);
    expect(result.props.children[4].props.children).toBe('error');
    expect(result.props.children[5].props.children[1].props.children.type).toBe(Spinner);
    expect(initialize).toHaveBeenCalled();
  });

  test('validate', () => {
    const initialize = jest.fn();
    const props = { ...PersonalDataReview.defaultProps, initialize };

    useParams.mockImplementation(() => ({ page: 'ktp', subpage: 'edit' }));

    const result = PersonalDataReview(props);

    const v1 = result.props.children[3].props.validate('tes');
    expect(v1).toBe('');

    const v2 = result.props.children[3].props.validate('');
    expect(v2).toBe('Harus diisi');
  });
});
