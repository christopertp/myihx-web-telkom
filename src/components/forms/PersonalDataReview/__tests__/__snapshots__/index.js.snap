// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/forms/PersonalDataReview render 1`] = `
<form
  onSubmit={[Function]}
>
  <Field
    component={[Function]}
    inputProps={
      Object {
        "label": "E-KTP NUMBER",
        "maxLength": 16,
        "placeholder": "Enter e-KTP number",
      }
    }
    name="idNumber"
  />
  <Field
    component={[Function]}
    inputProps={
      Object {
        "label": "NAME",
        "placeholder": "Enter name according to e-KTP",
      }
    }
    name="name"
  />
  <Field
    component={[Function]}
    inputProps={
      Object {
        "label": "MOTHER'S MAIDEN NAME",
        "placeholder": "Enter mother's maiden name",
      }
    }
    name="motherName"
  />
  <div>
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="bordered"
    >
      Reupload
    </Button>
    <Button
      className=""
      disabled={false}
      onClick={null}
      to=""
      type="submit"
      variant="red"
    >
      Confirm
    </Button>
  </div>
</form>
`;
